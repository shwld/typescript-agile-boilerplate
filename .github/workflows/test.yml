name: test

on:
  push:
    branches: ["main"]
  pull_request:
    types: [opened, synchronize]

jobs:
  build:
    name: Test
    timeout-minutes: 15
    runs-on: ${{ matrix.os }}
    # To use Remote Caching, uncomment the next lines and follow the steps below.
    # env:
    #  TURBO_TOKEN: ${{ secrets.TURBO_TOKEN }}
    #  TURBO_TEAM: ${{ secrets.TURBO_TEAM }}
    strategy:
      matrix:
        # os: [ubuntu-latest, macos-latest]
        os: [ubuntu-latest]
    services:
      # Label used to access the service container
      postgres:
        # Docker Hub image
        image: postgres
        # Provide the password for postgres
        env:
          POSTGRES_USER: postgres
          POSTGRES_PASSWORD: password
          POSTGRES_DB: app_test
        # Set health checks to wait until postgres has started
        options: >-
          --health-cmd pg_isready
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5

    steps:
      - name: Check out code
        uses: actions/checkout@v2
        with:
          fetch-depth: 2

      - name: Setup Node.js environment
        uses: actions/setup-node@v2
        with:
          node-version: 16
          cache: 'yarn'

      - name: Setup Test
        run: |
          echo 'DATABASE_URL="postgresql://postgres:password@postgres:5432/app_test?schema=public"' > ./infrastructures/db/.env
          echo 'DATABASE_URL="postgresql://postgres:password@postgres:5432/app_test?schema=public"' > ./use-cases/graphql-resolvers/.env
          yarn db:migrate

      - name: Install dependencies
        run: yarn

      - name: Test
        run: yarn test
