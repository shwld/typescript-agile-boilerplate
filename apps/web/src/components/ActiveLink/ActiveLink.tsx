// REF: https://github.com/vercel/next.js/blob/canary/examples/active-class-name/components/ActiveLink.tsx
import { FC } from 'react';
import { useRouter } from 'next/router';
import Link, { LinkProps } from 'next/link';
import React, { useState, useEffect, ReactElement, Children } from 'react';

type ActiveLinkProps = LinkProps & {
  children: ReactElement;
};

const ActiveLink: FC<ActiveLinkProps> = ({ children, ...props }) => {
  const { asPath, isReady } = useRouter();
  const child = Children.only(children);
  const [isActive, setActive] = useState(false);

  useEffect(() => {
    // Check if the router fields are updated client-side
    if (isReady) {
      // Dynamic route will be matched via props.as
      // Static route will be matched via props.href
      const linkPathname = new URL(
        (props.as || props.href) as string,
        location.href
      ).pathname;

      // Using URL().pathname to get rid of query and hash
      const activePathname = new URL(asPath, location.href).pathname;
      setActive(linkPathname === activePathname);
    }
  }, [asPath, isReady, props.as, props.href]);

  return (
    <Link {...props}>
      {React.cloneElement(child, {
        isActive,
      })}
    </Link>
  );
};

export default ActiveLink;
