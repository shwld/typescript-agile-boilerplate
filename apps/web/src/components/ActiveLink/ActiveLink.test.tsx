import { ActiveLink } from './ActiveLink';
import { render } from '@testing-library/react';
import { MockedRouterProvider } from '~/test/MockedRouterProvider';

describe('ActiveLink', () => {
  const renderComponent = () => {
    const renderResult = render(
      <MockedRouterProvider>
        <ActiveLink href="">
          <>test</>
        </ActiveLink>
      </MockedRouterProvider>
    );
    return renderResult;
  };
  test('Snapshot', () => {
    expect(renderComponent().asFragment()).toMatchSnapshot();
  });
  test('success', () => {
    const { getByText } = renderComponent();
    expect(getByText('test')).toBeTruthy();
  });
});
