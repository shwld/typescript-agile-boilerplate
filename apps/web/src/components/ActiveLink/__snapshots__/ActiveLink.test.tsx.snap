// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActiveLink Snapshot 1`] = `
<DocumentFragment>
  test
</DocumentFragment>
`;
